charmcraft/config.py

```diff
@@ -105,6 +114,73 @@ def from_dict(cls, source):
         return cls(**source)
 
 
+@attr.s(kw_only=True, frozen=True)
+class BaseConfig:
+    """Configuration for all Charmhub related options."""
+
+    name = attr.ib(validator=attr.validators.instance_of(str))
+    channel = attr.ib(validator=attr.validators.instance_of(str))
+    architectures = attr.ib(default=attr.Factory(list), validator=attr.validators.deep_iterable(
+        member_validator=attr.validators.instance_of(str),
+        iterable_validator=attr.validators.instance_of(list)))
+
+    @classmethod
+    def from_dict(cls, source):
+        """Build from a raw dict."""
+        # Ensure channel is treated as string if interpreted as number, e.g. 20.04.
```

Comment:
But

```
>>> str(20.10)
'20.1'
```

I propose to just error out if it's a float.

Reply:
Yikes, good point.  I was wondering if it's worth moving to strictyaml.load() to get some better behavior out of the box and avoid this kind of issue?

Reply:
`strictyaml` looks nice, but then everything will be read as strings, unless you give it a schema, and I don't like having the schema twice.

But IMO there's no need for that! Our schema already handles this detail, in the sense that it will error out with a clear message: `the 'channel' field must be a string: got 'float'.`. And the person writing the YAML can easily fix it. 

Remember our users are developers...